include/avif/avif.h

```diff
@@ -824,7 +824,7 @@ typedef struct avifImage
 
     // Other properties attached to this image item (primary or gainmap).
     // At decoding: Forwarded here as opaque byte sequences by the avifDecoder.
-    // At encoding: Ignored.
+    // At encoding: Set using avifImageAddOpaqueProperty() or avifImageAddUUIDProperty() and written by the avifEncoder.
```

Comment:
If the order of properties is important, we should document that these properties are written in the order in which they are added to the image, and after all the other properties generated by avifEncoder itself.

Reply:
Partly fixed. I've noted the part about the order of addition.

In doing this, I realised that there is a problem with doing it after "all the other properties". ISO/IEC 23008-12:2022 Section 6.5 requires that descriptive properties come before transformative properties. So adding at the end might or might not work depending on whether the reader actually enforces the part about ignoring descriptive properties after the first transformative or unrecognized property, whichever is earlier; and whether there are any transformative properties.

Separate issue here: `a1lx` might not be written in the  correct order - not sure I'm completely following that part of the code though.

Options:
1. Document that only descriptive properties are allowed, and move the addition up to before the transformative properties.
2. Use different structures for descriptive and transformative properties, and document they are added at the end of the descriptive and transformative properties (respectively) added by avifEncoder.

I lean to Option 1. 

Suggestions on other options? Preferences on implementation?

Reply:
> In doing this, I realised that there is a problem with doing it after "all the other properties". 

Good catch!

> ISO/IEC 23008-12:2022 Section 6.5 requires that descriptive properties come before transformative properties. So adding at the end might or might not work depending on whether the reader actually enforces the part about ignoring descriptive properties after the first transformative or unrecognized property, whichever is earlier; and whether there are any transformative properties.

It says "readers shall ignore". We should not care about non-compliant implementations, but in practice I agree it is best to make sure it works everywhere. libavif probably does not check that.  
Since these properties are descriptive and should not impact the output sample values, I wonder what is the difference between ignoring them and, well, not ignoring them.

> Separate issue here: `a1lx` might not be written in the correct order - not sure I'm completely following that part of the code though.

I think https://github.com/AOMediaCodec/libavif/pull/2524 would fix that.

> Options:
> 
> 1. Document that only descriptive properties are allowed, and move the addition up to before the transformative properties.
> 2. Use different structures for descriptive and transformative properties, and document they are added at the end of the descriptive and transformative properties (respectively) added by avifEncoder.
> 
> I lean to Option 1.
> 
> Suggestions on other options? Preferences on implementation?

Assuming we go with https://github.com/AOMediaCodec/libavif/pull/2524:
- The descriptive custom props impl can be inserted in write.c between the libavif-handled descriptive properties and the libavif-handled transformative properties.
- The public API can then allow both descriptive and transformative custom props, as long as they are in this order.
- Note that the spec says "Writers should arrange the descriptive properties specified in 6.5 prior to any other properties", so HEIF props come first, before non-HEIF descriptive props. 
  - This may be too much of a detail for the user to respect. I would avoid mentioning that in avif.h.
  - `a1lx` is specified in [av1-avif](https://aomediacodec.github.io/av1-avif/), not in HEIF. So if a user adds a HEIF property that is not handled by libavif, and `a1lx` is written, this constraint will not be fulfilled. I think we can ignore this constraint because it is fairly niche and only a "should". A comment explaining this part would be nice (not necessarily in the header though).

Reply:
> * The descriptive custom props impl can be inserted in write.c between the libavif-handled descriptive properties and the libavif-handled transformative properties.
> 
> * The public API can then allow both descriptive and transformative custom props, as long as they are in this order.

That isn't going to work for all cases, because `ispe` has to be the first transformative property (see ISO/IEC 23008-12:2022 Section 6.5.3.1).

Reply:
It would probably be easier if #2524 went in before trying to update this PR to reflect that.

Reply:
Done!

Reply:
I also filed https://github.com/AOMediaCodec/libavif/issues/2525.

Reply:
Thanks. I've rebased against main branch, which predictably caused merge conflicts. I've moved up the opaque property writing to be before transformative properties (as suggested), and added documentation about the ordering.